lib/internal/modules/esm/loader.js

```diff
@@ -168,100 +212,250 @@ class Loader {
     const job = new ModuleJob(this, url, evalInstance, false, false);
     this.moduleMap.set(url, job);
     const { module } = await job.run();
+
     return {
       namespace: module.getNamespace(),
     };
   }
 
-  async import(specifier, parent) {
-    const job = await this.getModuleJob(specifier, parent);
-    const { module } = await job.run();
-    return module.getNamespace();
+  async getModuleJob(specifier, parentURL) {
+    const { format, url } = await this.resolve(specifier, parentURL);
+    let job = this.moduleMap.get(url);
+    // CommonJS will set functions for lazy job evaluation.
+    if (typeof job === 'function') this.moduleMap.set(url, job = job());
+
+    if (job !== undefined) return job;
+
+    const moduleProvider = async (url, isMain) => {
+      const { format: finalFormat, source } = await this.load(url, { format });
```

Comment:
We might want to restrict `format` and `source` to be own properties of the loader to avoid prototype pollution. Or we could leave it to hook authors to protect their load hooks from this. I don't have a strong opinion on either, but I think we should document it whatever we choose.